### PR TITLE
More pipeline cleanups

### PR DESCRIPTION
**Pull request by dabrahams**
I apologize for the mess of these commits.

### CODE REVIEW

File: Sources/CodeGen/CXX/CXXTranspiler.swift [hunk 1]
```diff
@@ -6,6 +6,8 @@ import Utils
 /// The conversion of typed Val module AST into corresponding C++ AST.
 public struct CXXTranspiler {
 
+  public static let coreLibModuleName = "ValStdLib"
```
Comment (kyouko-taiga) [inline Sources/CodeGen/CXX/CXXTranspiler.swift:9]: I'm surprised to see this after #399 was opened. Just want to make sure this line is intentional.

File: Sources/ValCommand/ValCommand.swift [hunk 2]
```diff
@@ -135,103 +135,65 @@ public struct ValCommand: ParsableCommand {
...
     if outputType == .rawAST {
-      let url = outputURL ?? URL(fileURLWithPath: productName + ".ast.json")
-      let encoder = JSONEncoder().forAST
-      try encoder.encode(ast).write(to: url, options: .atomic)
+      try write(ast, to: astFile(productName))
       return
     }
 
-    let typedProgram = try TypedProgram(
+    let p = try TypedProgram(
```
Comment (kyouko-taiga) [inline Sources/ValCommand/ValCommand.swift:150]: I feel the scope of this variable is a bit too large for it to be a single-letter identifier.

I can't think of a much better name than `typedProgram`. That's the role this variable has here: the program we're compiling, typed.

File: Sources/ValCommand/ValCommand.swift [hunk 3]
```diff
@@ -341,4 +302,49 @@ public struct ValCommand: ParsableCommand {
   /// A map from executable name to path of the named binary.
   private static var executableLocationCache: [String: String] = [:]
 
+  /// Wrotes a textual descriptioni of `input` to the given `output` file.
+  func write(_ input: AST, to output: URL) throws {
+    let encoder = JSONEncoder().forAST
+    try encoder.encode(input).write(to: output, options: .atomic)
+  }
+
+  /// Given the product name (whatever that means), returns the file to write when "raw-ast" is
```
Comment (kyouko-taiga) [inline Sources/ValCommand/ValCommand.swift:311]: > (whatever that means)

Hahahaha.
Well, I think we should discuss this.

File: Sources/ValCommand/ValCommand.swift [hunk 4]
```diff
@@ -341,4 +302,49 @@ public struct ValCommand: ParsableCommand {
   /// A map from executable name to path of the named binary.
   private static var executableLocationCache: [String: String] = [:]
 
+  /// Wrotes a textual descriptioni of `input` to the given `output` file.
```
Comment (kyouko-taiga) [inline Sources/ValCommand/ValCommand.swift:305]:
```suggestion
  /// Writes a textual descriptioni of `input` to the given `output` file.
```

File: Sources/ValCommand/ValCommand.swift [hunk 5]
```diff
@@ -341,4 +302,49 @@ public struct ValCommand: ParsableCommand {
...
+    outputURL?.deletingPathExtension() ?? URL(fileURLWithPath: productName)
+  }
+
+}
+
+extension TypedProgram {
+
+  typealias CXX = (syntax: CXXModule, text: TranslationUnitCode)
+
+  func cxx(_ m: ModuleDecl.Typed) -> CXX {
```
Comment (kyouko-taiga) [inline Sources/ValCommand/ValCommand.swift:344]: No docs?

File: Sources/ValCommand/ValCommand.swift [hunk 6]
```diff
@@ -135,103 +135,65 @@ public struct ValCommand: ParsableCommand {
...
       return
     }
 
-    // *** C++ Transpiling ***
+    // C++
 
-    // Initialize the transpiler & code writer.
-    let transpiler = CXXTranspiler(typedProgram)
-    var codeWriter = CXXCodeWriter()
+    let cxxModules = (core: p.cxx(p.corelib!), source: p.cxx(p[sourceModule]))
```
Comment (lucteo) [inline Sources/ValCommand/ValCommand.swift:167]: neat!

File: Sources/ValCommand/ValCommand.swift [hunk 7]
```diff
@@ -341,4 +302,49 @@ public struct ValCommand: ParsableCommand {
...
+
+}
+
+extension TypedProgram {
+
+  typealias CXX = (syntax: CXXModule, text: TranslationUnitCode)
+
+  func cxx(_ m: ModuleDecl.Typed) -> CXX {
+    let x = CXXTranspiler(self).cxx(m)
+    var w = CXXCodeWriter()
```
Comment (lucteo) [inline Sources/ValCommand/ValCommand.swift:346]: Note: I'm also refactoring this, eliminating the need for CXXCodeWriter variable. See https://github.com/val-lang/val/commit/62d4ebba881d04c9a518d5cf346d61e9c9826b7e#diff-afa50b06bcd7e4dfe5601a33c9aa656568e9784f1f0699a4ef9fe2c683705d13

I can put my changes on top of this.

I'm starting to contemplate now if we should put this extension in CXXTranspiler itself, and remove the need to publicly use the class. (similar to what I've started to do with CXXCodeWriter)

File: Sources/ValCommand/ValCommand.swift [hunk 8]
```diff
@@ -341,4 +302,49 @@ public struct ValCommand: ParsableCommand {
...
+  /// "cpp" is selected as the output type.
+  private func sourceModuleCXXOutputBase(_ productName: String) -> URL {
+    outputURL?.deletingPathExtension() ?? URL(fileURLWithPath: productName)
+  }
+
+}
+
+extension TypedProgram {
+
+  typealias CXX = (syntax: CXXModule, text: TranslationUnitCode)
```
Comment (lucteo) [inline Sources/ValCommand/ValCommand.swift:342]: I don' think we use `syntax` anymore. If that's the case, then `CXX` == `TranslationUnitCode`.

Also, I would avoid using both `CXX` and `cxx` in the same context, as they can be easily misread (same shape, just the size of the letters are different)

**Review by lucteo — COMMENTED** [review]
LGTM

File: Sources/CodeGen/CXX/CXXTranspiler.swift (hunk 1 above)
Comment (dabrahams) [reply Sources/CodeGen/CXX/CXXTranspiler.swift:9]: Very.  #399 is about the identifiers used.  Other stuff is a separate issue.

File: Sources/ValCommand/ValCommand.swift (hunk 3 above)
Comment (dabrahams) [reply Sources/ValCommand/ValCommand.swift:311]: I think you should define it, but I'll be happy to talk to you about that if you want.

File: Sources/ValCommand/ValCommand.swift (hunk 8 above)
Comment (dabrahams) [reply Sources/ValCommand/ValCommand.swift:342]: `func index(after i: Index) -> Index` is a problem?

File: Sources/ValCommand/ValCommand.swift (hunk 7 above)
Comment (dabrahams) [reply Sources/ValCommand/ValCommand.swift:362]: Yes, the `struct` is not a publicly-useful abstraction.  If it should exist at all, it should be `fileprivate`.

File: Sources/ValCommand/ValCommand.swift (hunk 8 above)
Comment (dabrahams) [reply Sources/ValCommand/ValCommand.swift]: I renamed it `TypedProgram.CXXModule`

File: Sources/ValCommand/ValCommand.swift (hunk 5 above)
Comment (dabrahams) [reply Sources/ValCommand/ValCommand.swift]: fixed

File: Sources/ValCommand/ValCommand.swift (hunk 2 above)
Comment (dabrahams) [reply Sources/ValCommand/ValCommand.swift]: "The typed program" is not a role.  A role would be something more like “the source of the IR”

**Pull request merged**